python/lsst/afw/geom/span.cc

```diff
@@ -1,96 +1,98 @@
-/* 
+/*
  * LSST Data Management System
  * Copyright 2008-2016  AURA/LSST.
- * 
+ *
  * This product includes software developed by the
  * LSST Project (http://www.lsst.org/).
  *
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.
- * 
+ *
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
- * 
- * You should have received a copy of the LSST License Statement and 
- * the GNU General Public License along with this program.  If not, 
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
  * see <https://www.lsstcorp.org/LegalNotices/>.
  */
 
-#include <pybind11/pybind11.h>
-#include <pybind11/operators.h>
-#include <pybind11/stl.h>
+#include "pybind11/pybind11.h"
+#include "pybind11/stl.h"
 
 #include "lsst/afw/geom/Span.h"
 #include "lsst/afw/geom/SpanPixelIterator.h"
 
 namespace py = pybind11;
 
-using namespace lsst::afw::geom;
+namespace lsst { namespace afw { namespace geom { namespace {
+
+using PySpan = py::class_<Span, std::shared_ptr<Span>>;
 
 // A thin wrapper around SpanPixelIterator.
 // Unfortunately we cannot use py::make_iterator here, as we normally
 // should, because for some reason the return values then all refer
 // to the last element.
 class SpanIterator {
-    public:
-        SpanIterator(const Span &s) : _it{s.begin()}, _end{s.end()} {};
-        Point2I next() {
-            if (_it == _end) {
-                throw py::stop_iteration();
-            }
-            return *_it++;
-        };
-    private:
-        Span::Iterator _it;
-        Span::Iterator _end;
+public:
+    SpanIterator(const Span &s) : _it{s.begin()}, _end{s.end()} {};
+    Point2I next() {
+        if (_it == _end) {
+            throw py::stop_iteration();
+        }
+        return *_it++;
+    };
+private:
+    Span::Iterator _it;
+    Span::Iterator _end;
```

Comment:
There already is an upstream utility function for this: `pybind11::make_iterator`.  It does not work here, for reasons we do not understand (but may be related to the fact that `SpanPixelIterator` is not a legal `RandomAccessIterator` because it yields proxies).  This is [DM-7658](https://jira.lsstcorp.org/browse/DM-7658)